lib/dispense-generator.js

```diff
@@ -0,0 +1,80 @@
+const _ = require('lodash/fp')
+const actionEmitter = require('./action-emitter')
+const coinUtils = require('./coins/utils')
+
+const emit = event => actionEmitter.emit('dispenseGenerator', event)
+
+const dispenseGenerator = function* dispenseGenerator(notes, tx, txId) {
+  let result = []
+  
+  for (let index = 0; index < notes.length; index++) {
+    emit({ action: 'updateUI', current: index + 1, of: notes.length })
+    let yielded = yield emit({ action: 'dispenseBatch', notes: notes[index] })
+    if (yielded !== undefined) result.push(yielded)
+  }
+  
+  const bills = result.map(d => d.bills).reduce((p, c) => {
+    return [
+    {dispensed: p[0].dispensed + c[0].dispensed, rejected: p[0].rejected + c[0].rejected},
+    {dispensed: p[1].dispensed + c[1].dispensed, rejected: p[1].rejected + c[1].rejected}
+  ]})
+
+  fillInBills(tx, bills)
+  const dispenseConfirmed = fullDispense(tx)
+
+  if (dispenseConfirmed) emit({action: 'billDispenserDispensed'})
+    // update tx and keep track of the
+    // dispensed bills
+    // and the error code (if any)
+  emit({
```

Comment:
I'd prefer if the emitted object was defined separately and broken down into smaller parts to make it easier to read.